CS4001/4042 Assignment 1, Part B, Q4
---

Model degradation is a common issue faced when deploying machine learning models (including neural networks) in the real world. New data points could exhibit a different pattern from older data points due to factors such as changes in government policy or market sentiments. For instance, housing prices in Singapore have been increasing and the Singapore government has introduced 3 rounds of cooling measures over the past years (16 December 2021, 30 September 2022, 27 April 2023).

In such situations, the distribution of the new data points could differ from the original data distribution which the models were trained on. Recall that machine learning models often work with the assumption that the test distribution should be similar to train distribution. When this assumption is violated, model performance will be adversely impacted.  In the last part of this assignment, we will investigate to what extent model degradation has occurred.




---



---



Your co-investigators used a linear regression model to rapidly test out several combinations of train/test splits and shared with you their findings in a brief report attached in Appendix A below. You wish to investigate whether your deep learning model corroborates with their findings.

In [ ]:
!pip install alibi-detect

In [2]:
SEED = 42

import os
import torch

import random
random.seed(SEED)

import numpy as np
np.random.seed(SEED)

import pandas as pd
from sklearn.metrics import r2_score
from pytorch_tabular import TabularModel

from alibi_detect.cd import TabularDrift

> Evaluate your model from B1 on data from year 2022 and report the test R2.

In [3]:
df = pd.read_csv('hdb_price_prediction.csv')

df_eval_2022 = df[df['year'] == 2022]

tabular_model = TabularModel.load_from_checkpoint("hdb_regression")
pred_2022 = tabular_model.predict(df_eval_2022)

r2_score_2022 = r2_score(df_eval_2022['resale_price'], pred_2022['resale_price_prediction'])

print("R2 Score:", r2_score_2022)

2023-10-05 01:01:37,342 - {pytorch_tabular.tabular_model:129} - INFO - Experiment Tracking is turned off
2023-10-05 01:01:37,357 - {pytorch_tabular.tabular_model:268} - INFO - Preparing the Trainer
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\trainer\connectors\accelerator_connector.py:589: LightningDeprecationWarning: The Trainer argument `auto_select_gpus` has been deprecated in v1.9.0 and will be removed in v2.0.0. Please use the function `pytorch_lightning.accelerators.find_usable_cuda_devices` instead.
  rank_zero_deprecation(
Trainer already configured with model summary callbacks: [<class 'pytorch_lightning.callbacks.rich_model_summary.RichModelSummary'>]. Skipping setting a default `ModelSummary` callback.
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs


Output()

R2 Score: 0.4928212504557957


> Evaluate your model from B1 on data from year 2023 and report the test R2.

In [4]:
# TODO: Enter your code here
df = pd.read_csv('hdb_price_prediction.csv')

df_eval_2023 = df[df['year'] == 2023]

tabular_model = TabularModel.load_from_checkpoint("hdb_regression")
pred_2023 = tabular_model.predict(df_eval_2023)

r2_score_2023 = r2_score(df_eval_2023['resale_price'], pred_2023['resale_price_prediction'])

print("R2 Score:", r2_score_2023)

2023-10-05 01:01:43,898 - {pytorch_tabular.tabular_model:129} - INFO - Experiment Tracking is turned off
2023-10-05 01:01:43,913 - {pytorch_tabular.tabular_model:268} - INFO - Preparing the Trainer
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\trainer\connectors\accelerator_connector.py:589: LightningDeprecationWarning: The Trainer argument `auto_select_gpus` has been deprecated in v1.9.0 and will be removed in v2.0.0. Please use the function `pytorch_lightning.accelerators.find_usable_cuda_devices` instead.
  rank_zero_deprecation(
Trainer already configured with model summary callbacks: [<class 'pytorch_lightning.callbacks.rich_model_summary.RichModelSummary'>]. Skipping setting a default `ModelSummary` callback.
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs


Output()

R2 Score: 0.2228574322089607


> Did model degradation occur for the deep learning model?


Yes model degredation did occur. As the model was train on <= 2019 data, there might have been real world conditions affecting the data features of the model in the year 2023.



---



---



Model degradation could be caused by [various data distribution shifts](https://huyenchip.com/2022/02/07/data-distribution-shifts-and-monitoring.html#data-shift-types): covariate shift (features), label shift and/or concept drift (altered relationship between features and labels).
There are various conflicting terminologies in the [literature](https://www.sciencedirect.com/science/article/pii/S0950705122002854#tbl1). Let’s stick to this reference for this assignment.

> Using the **Alibi Detect** library, apply the **TabularDrift** function with the training data (year 2019 and before) used as the reference and **detect which features have drifted** in the 2023 test dataset. Before running the statistical tests, ensure you **sample 1000 data points** each from the train and test data. Do not use the whole train/test data. (Hint: use this example as a guide https://docs.seldon.io/projects/alibi-detect/en/stable/examples/cd_chi2ks_adult.html)


In [27]:
# TODO: Enter your code here
num_col_names = ['dist_to_nearest_stn', 'dist_to_dhoby', 'degree_centrality', 'eigenvector_centrality', 'remaining_lease_years', 'floor_area_sqm']
cat_col_names = ['month', 'town', 'flat_model_type', 'storey_range']

df_reference = df[df['year'] <= 2019]
y_df_reference = df_reference['resale_price']

df_test = df[df['year'] == 2023]
y_df_test = df_test['resale_price']

n_ref = 1000
n_test = 1000
df_reference = df_reference[:n_ref][cat_col_names+num_col_names] 
df_test = df_test[:n_test][cat_col_names+num_col_names] 

category_map = {}
features = cat_col_names+num_col_names
for key, value in enumerate(features):
    if value in cat_col_names:
        category_map[key] = df[value].unique().tolist()

categories_per_feature = {f: None for f in category_map.keys()}

cd = TabularDrift(df_reference.values, p_val=.05, categories_per_feature=categories_per_feature)

preds = cd.predict(df_test.values)
labels = ['No!', 'Yes!']
print('Drift? {}'.format(labels[preds['data']['is_drift']]))

threshold = preds['data']['threshold']
print("Threshold:", threshold)

Drift? Yes!
Threshold: 0.005


In [29]:
feature_names = df_test.columns.values
fpreds = cd.predict(df_test.values, drift_type='feature')

drifted_feature_list = []
for f in range(cd.n_features):
    stat = 'Chi2' if f in list(categories_per_feature.keys()) else 'K-S'
    fname = feature_names[f]
    is_drift = fpreds['data']['is_drift'][f]
    stat_val, p_val = fpreds['data']['distance'][f], fpreds['data']['p_val'][f]
    print(f'{fname} -- Drift? {labels[is_drift]} -- {stat} {stat_val:.3f} -- p-value {p_val:.3f}')
    
    if is_drift:
        drifted_feature_list.append(fname)

month -- Drift? No! -- Chi2 0.000 -- p-value 1.000
town -- Drift? Yes! -- Chi2 667.474 -- p-value 0.000
flat_model_type -- Drift? Yes! -- Chi2 77.586 -- p-value 0.000
storey_range -- Drift? Yes! -- Chi2 38.800 -- p-value 0.001
dist_to_nearest_stn -- Drift? No! -- K-S 0.055 -- p-value 0.094
dist_to_dhoby -- Drift? Yes! -- K-S 0.218 -- p-value 0.000
degree_centrality -- Drift? No! -- K-S 0.029 -- p-value 0.783
eigenvector_centrality -- Drift? Yes! -- K-S 0.195 -- p-value 0.000
remaining_lease_years -- Drift? Yes! -- K-S 0.271 -- p-value 0.000
floor_area_sqm -- Drift? Yes! -- K-S 0.134 -- p-value 0.000


> Assuming that the flurry of housing measures have made an impact on the relationship between all the features and resale_price (i.e. P(Y|X) changes), which type of data distribution shift possibly led to model degradation?


Concept drift

> From your analysis via TabularDrift, which features contribute to this shift?


In [30]:
for features in drifted_feature_list:
    print(features)

town
flat_model_type
storey_range
dist_to_dhoby
eigenvector_centrality
remaining_lease_years
floor_area_sqm


Features contributing to this shift include:
- town
- flat_model_type
- storey_range
- dist_to_dhoby
- eigenvector_centrality
- remaining_lease_years
- floor_area_sqm

> Suggest 1 way to address model degradation and implement it, showing improved test R2 for year 2023.


Update (Retrain) the model on new data whenever new data is available.

In [31]:
# TODO: Enter your code here
tabular_model.fit(train=df[df['year'] <= 2022])

Global seed set to 42
2023-10-05 01:47:47,070 - {pytorch_tabular.tabular_model:473} - INFO - Preparing the DataLoaders
2023-10-05 01:47:47,080 - {pytorch_tabular.tabular_datamodule:290} - INFO - Setting up the datamodule for regression task
2023-10-05 01:47:47,294 - {pytorch_tabular.tabular_model:521} - INFO - Preparing the Model: CategoryEmbeddingModel
2023-10-05 01:47:47,327 - {pytorch_tabular.tabular_model:268} - INFO - Preparing the Trainer
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\trainer\connectors\accelerator_connector.py:589: LightningDeprecationWarning: The Trainer argument `auto_select_gpus` has been deprecated in v1.9.0 and will be removed in v2.0.0. Please use the function `pytorch_lightning.accelerators.find_usable_cuda_devices` instead.
  rank_zero_deprecation(
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
2023-10-05 01:47:47,452 - {p

Finding best initial lr:   0%|          | 0/100 [00:00<?, ?it/s]

c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: You called `self.log('train_loss', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: You called `self.log('train_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(
`Trainer.fit` stopped: `max_steps=100` reached.
Learning rate set to 0.5754399373371567
Restoring states from the checkpoint path at d:\micha\Michael\GitHub\Michael_Lee\CZ4042_lab\Assignment\.lr_find_728aa1e8-c29c-462e-85f3-625f6cab7877.ckpt
Restored all states from the checkpoint file at d:\micha\Michael\GitHub\Michael_Lee\CZ4042_lab\Assignment\.lr_find_728aa1e8-c29c-462e-85f3-625f6cab7877.ckpt
2023-10-05 01:47:51,699 - {pytorch_

┏━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃   ┃ Name             ┃ Type                      ┃ Params ┃
┡━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│ 0 │ _backbone        │ CategoryEmbeddingBackbone │  3.0 K │
│ 1 │ _embedding_layer │ Embedding1dLayer          │  1.6 K │
│ 2 │ head             │ LinearHead                │     51 │
│ 3 │ loss             │ MSELoss                   │      0 │
└───┴──────────────────┴───────────────────────────┴────────┘

Trainable params: 4.6 K                                                                                            
Non-trainable params: 0                                                                                            
Total params: 4.6 K                                                                                                
Total estimated model params size (MB): 0

Output()

2023-10-05 01:50:19,116 - {pytorch_tabular.tabular_model:584} - INFO - Training the model completed
2023-10-05 01:50:19,117 - {pytorch_tabular.tabular_model:1258} - INFO - Loading the best model
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\utilities\cloud_io.py:33: LightningDeprecationWarning: `pytorch_lightning.utilities.cloud_io.get_filesystem` has been deprecated in v1.8.0 and will be removed in v2.0.0. Please use `lightning_fabric.utilities.cloud_io.get_filesystem` instead.
  rank_zero_deprecation(


In [32]:
pred_2023 = tabular_model.predict(df_eval_2023)
r2_score_2023 = r2_score(df_eval_2023['resale_price'], pred_2023['resale_price_prediction'])

print("R2 Score:", r2_score_2023)

Output()

R2 Score: 0.5226147479107343


R2 score after retraining the model with newer data has increased from 0.223 to 0.523. An improvment is observed.

### Appendix A



Here are our results from a linear regression model. We used StandardScaler for continuous variables and OneHotEncoder for categorical variables.

While 2021 data can be predicted well, test R2 dropped rapidly for 2022 and 2023.

| Training set | Test set | Test R2 |
|--------------|----------|---------|
| Year <= 2020 | 2021     | 0.76    |
| Year <= 2020 | **2022**     | 0.41    |
| Year <= 2020 | **2023**     | **0.10**   |



Similarly, a model trained on 2017 data can predict 2018-2021 well (with slight degradation in performance for 2021), but drops drastically in 2022 and 2023.

| Training set | Test set | Test R2 |
|--------------|----------|---------|
| 2017         | 2018     | 0.90    |
|              | 2019     | 0.89    |
|              | 2020     | 0.87    |
|              | 2021     | 0.72    |
|              | **2022**     | **0.37**    |
|              | **2023**     | **0.09**    |

With the test set fixed at year 2021, training on data from 2017-2020 still works well on the test data, with minimal degradation. Training sets closer to year 2021 generally do better.

| Training set | Test set | Test R2 |
|--------------|----------|---------|
| 2020         | 2021     | 0.81    |
| 2019         | 2021     | 0.75    |
| 2018         | 2021     | 0.73    |
| 2017         | 2021     | 0.72    |